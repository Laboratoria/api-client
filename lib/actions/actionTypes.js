export default apiCalls => apiCalls.reduce((acc, apiCall) => {
  const { type } = apiCall;
  return {
    ...acc,
    [`API_${type}`]: `API_${type}`,
    [`API_${type}_SUCCESS`]: `API_${type}_SUCCESS`,
    [`API_${type}_FAILURE`]: `API_${type}_FAILURE`
  };
}, {});
