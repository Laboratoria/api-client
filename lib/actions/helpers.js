import { API_REQUEST } from './types';

export const laboratoriaAPIAction = ({
  type,
  url,
  method = 'GET',
  headers,
  params,
  data,
  anonymous,
  key,
  expiration,
  namespace = 'default',
}) => ({
  type: `${API_REQUEST}${type}`,
  payload: {
    url,
    method,
    headers,
    params,
    data,
  },
  meta: {
    anonymous,
    expiration,
    key,
    suffix: type,
    namespace,
  },
});
