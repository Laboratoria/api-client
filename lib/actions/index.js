import Api from '../api';
import APIClient from '../APIClient';

export default dispatch => (apiCallName, params = {}, queryData, bodyData) => {
  const apiCall = APIClient.apiActions().find(apiCallData => apiCallData.type === apiCallName);
  let url = apiCall.relativeUrl;
  Object.entries(params).forEach((param) => {
    url = url.replace(`:${param[0]}`, param[1]);
  });
  Api.apiCall(url, apiCall.method, queryData, bodyData)
    .then(response => dispatch({
      type: `API_${apiCall.type}_SUCCESS`,
      payload: response,
    }))
    .catch(error => dispatch({
      type: `API_${apiCall.type}_FAILURE`,
      payload: error,
    }));

  return dispatch({
    type: `API_${apiCall.type}`,
  }); // Mark this call as loading...
};
