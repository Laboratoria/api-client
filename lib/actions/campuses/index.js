import { laboratoriaAPIAction } from '../helpers';

export const getCampuses = ({
  ...rest
} = {}) => laboratoriaAPIAction({
  type: 'CAMPUSES',
  url: '/campuses',
  method: 'GET',
  key: 'campuses',
  expiration: 1000 * 60 * 60,
  ...rest,
});
