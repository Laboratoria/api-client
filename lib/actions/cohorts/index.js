import { laboratoriaAPIAction } from '../helpers';

export const getCohorts = ({
  campus,
  program,
  track,
  page = 1,
  limit = 100,
  ...rest
} = {}) => laboratoriaAPIAction({
  type: 'COHORTS',
  url: '/cohorts',
  method: 'GET',
  key: 'cohorts',
  params: {
    campus,
    program,
    track,
    page,
    limit,
  },
  expiration: 1000 * 60 * 60,
  ...rest,
});
