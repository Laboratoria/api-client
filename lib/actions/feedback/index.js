import { laboratoriaAPIAction } from '../helpers';

export const getFeedbacks = ({
  user,
  cohortId,
  projectId,
  ...rest
}) => laboratoriaAPIAction({
  type: 'FEEDBACKS',
  url: `/users/${user}/cohorts/${cohortId}/projects/${projectId}/feedbacks`,
  method: 'GET',
  key: `user/${user}/feedbacks`,
  ...rest,
});

export const getFeedbacksByType = ({
  user,
  cohortId,
  projectId,
  reviewerSurvey,
  ...rest
}) => laboratoriaAPIAction({
  type: 'FEEDBACK_REVIEWER_SURVEY',
  url: `/users/${user}/cohorts/${cohortId}/projects/${projectId}/feedbacks/${reviewerSurvey}`,
  method: 'GET',
  key: `user/${user}/feedbacks/${reviewerSurvey}`,
  ...rest,
});

export const addFeedback = ({
  user,
  cohortId,
  projectId,
  data,
  ...rest
}) => laboratoriaAPIAction({
  type: 'FEEDBACK_ADD',
  url: `/users/${user}/cohorts/${cohortId}/projects/${projectId}/feedbacks`,
  method: 'POST',
  data,
  key: 'feedback-add',
  ...rest,
});

export const updateFeedback = ({
  id,
  data,
  ...rest
}) => laboratoriaAPIAction({
  type: 'FEEDBACK_UPDATE',
  url: `/feedbacks/${id}`,
  method: 'PUT',
  data,
  key: 'feedback-update',
  ...rest,
});

export const deleteFeedback = ({
  id,
  ...rest
}) => laboratoriaAPIAction({
  type: 'FEEDBACK_DELETE',
  url: `/feedbacks/${id}`,
  method: 'DELETE',
  key: 'feedback-delete',
  ...rest,
});

export const sendFeedback = ({
  user,
  cohortId,
  projectId,
  reviewerSurvey,
  ...rest
}) => laboratoriaAPIAction({
  type: 'FEEDBACK_SEND',
  url: `/users/${user}/cohorts/${cohortId}/projects/${projectId}/feedbacks/${reviewerSurvey}/_send`,
  method: 'POST',
  key: 'feedback-send',
  ...rest,
});
