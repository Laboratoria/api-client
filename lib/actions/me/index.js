import { laboratoriaAPIAction } from '../helpers';

export const updateProfile = ({
  data,
  ...rest
}) => laboratoriaAPIAction({
  type: 'UPDATE_MY_PROFILE',
  url: '/me/profile',
  method: 'PATCH',
  data,
  key: 'update-my-profile',
  ...rest,
});

export const createMyProject = ({
  data,
  ...rest
}) => laboratoriaAPIAction({
  type: 'CREATE_MY_PROJECT',
  url: '/me/profile/projects',
  method: 'POST',
  data,
  key: 'create-my-project',
  ...rest,
});

export const updateMyProject = ({
  id,
  data,
  ...rest
}) => laboratoriaAPIAction({
  type: 'UPDATE_MY_PROJECT',
  url: `/me/profile/projects/${id}`,
  method: 'PATCH',
  data,
  key: 'update-my-project',
  ...rest,
});

export const deleteMyProject = ({
  id,
  ...rest
}) => laboratoriaAPIAction({
  type: 'DELETE_MY_PROJECT',
  url: `/me/profile/projects/${id}`,
  method: 'DELETE',
  key: 'delete-my-project',
  ...rest,
});
