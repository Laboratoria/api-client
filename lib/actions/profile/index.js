import { laboratoriaAPIAction } from '../helpers';

export const getComments = ({ user, ...rest }) => laboratoriaAPIAction({
  type: 'PROFILE_COMMENTS',
  url: `/users/${user}/profile/comments`,
  method: 'GET',
  key: `user/${user}/profile/comments`,
  expiration: 1000 * 60 * 5,
  ...rest,
});
