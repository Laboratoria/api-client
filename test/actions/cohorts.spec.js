import { getCohorts } from '../../lib/actions/cohorts';

describe('getCohorts', () => {
  it('should be a function', () => {
    expect(typeof getCohorts).toBe('function');
  });

  it('should create an appropriate action', () => {
    expect(getCohorts()).toMatchSnapshot();
  });
});
