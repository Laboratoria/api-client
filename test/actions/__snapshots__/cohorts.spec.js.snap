// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCohort should create an appropriate action 1`] = `
Object {
  "meta": Object {
    "anonymous": undefined,
    "expiration": undefined,
    "key": "cohorts/someone",
    "namespace": "default",
    "suffix": "COHORT",
  },
  "payload": Object {
    "data": undefined,
    "headers": undefined,
    "method": "GET",
    "params": undefined,
    "url": "/cohorts/someone",
  },
  "type": "@@api-client/API_REQUEST/COHORT",
}
`;
